### ISSUE TITLE
Timeout Error in Final Iterations of Last Epoch Due to Process Desynchronization

### ISSUE THREAD

**Issue by zhangshuoneu** [facebookresearch/vggt#301]
Hi Jianyuan,

I've noticed that my training process generally runs smoothly, but I consistently encounter a timeout error during the final iterations of the last epoch.

It appears to be a synchronization issue among the multiple GPU processes. Some processes seem to finish processing their data and stop updating gradients, while others are still training. Consequently, the All-Reduce operation hangs because it cannot collect gradients from all processes, which ultimately leads to a timeout.

I was wondering if you have any guidance or suggestions on how to address this issue.

Thanks for your time and help.

[rank15]:[E726 13:23:24.941166255 ProcessGroupNCCL.cpp:616] [Rank 15] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=1201955, OpType=ALLREDUCE, NumelIn=7020002, NumelOut=7020002, Timeout(ms)=1800000) ran for 1800066 milliseconds before timing out.
[rank15]:[E726 13:23:24.941384382 ProcessGroupNCCL.cpp:1785] [PG ID 0 PG GUID 0(default_pg) Rank 15] Exception (either an error or timeout) detected by watchdog at work: 1201955, last enqueued NCCL work: 1202080, last completed NCCL work: 1201954.
[rank15]:[E726 13:24:01.757588896 ProcessGroupNCCL.cpp:1834] [PG ID 0 PG GUID 0(default_pg) Rank 15] Timeout at NCCL work: 1201955, last enqueued NCCL work: 1202080, last completed NCCL work: 1201954.
[rank15]:[E726 13:24:01.757617464 ProcessGroupNCCL.cpp:630] [Rank 15] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[rank15]:[E726 13:24:01.757626497 ProcessGroupNCCL.cpp:636] [Rank 15] To avoid data inconsistency, we are taking the entire process down.
[rank15]:[E726 13:24:01.758927089 ProcessGroupNCCL.cpp:1595] [PG ID 0 PG GUID 0(default_pg) Rank 15] Process group watchdog thread terminated with exception: [Rank 15] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=1201955, OpType=ALLREDUCE, NumelIn=7020002, NumelOut=7020002, Timeout(ms)=1800000) ran for 1800066 milliseconds before timing out.
Exception raised from checkTimeout at ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:618 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x96 (0x7fce3c36c446 in /home/suxin/anaconda3/envs/slam3r/lib/python3.11/site-packages/torch/lib/libc10.so)
frame #1: c10d::ProcessGroupNCCL::WorkNCCL::checkTimeout(std::optional<std::chrono::duration<long, std::ratio<1l, 1000l> > >) + 0x282 (0x7fcdf2219772 in /home/suxin/anaconda3/envs/slam3r/lib/python3.11/site-packages/torch/lib/libtorch_cuda.so)
frame #2: c10d::ProcessGroupNCCL::watchdogHandler() + 0x233 (0x7fcdf2220bb3 in /home/suxin/anaconda3/envs/slam3r/lib/python3.11/site-packages/torch/lib/libtorch_cuda.so)
frame #3: c10d::ProcessGroupNCCL::ncclCommWatchdog() + 0x14d (0x7fcdf222261d in /home/suxin/anaconda3/envs/slam3r/lib/python3.11/site-packages/torch/lib/libtorch_cuda.so)
frame #4: <unknown function> + 0x145c0 (0x7fce3c78a5c0 in /home/suxin/anaconda3/envs/slam3r/lib/python3.11/site-packages/torch/lib/libtorch.so)
frame #5: <unknown function> + 0x94ac3 (0x7fce3d870ac3 in /lib/x86_64-linux-gnu/libc.so.6)
frame #6: <unknown function> + 0x126850 (0x7fce3d902850 in /lib/x86_64-linux-gnu/libc.so.6)

**Comment by jytime** [comment]
Hi in my case usually this means some files are corrupted and your dataloader worker cannot read it successfully, which will then actually hang

**Comment by zhangshuoneu** [comment]
Hi Jianyuan,

Thanks for the reply and the suggestion about corrupted data files.

What I find puzzling is that the training process runs perfectly for nearly the entire duration. The timeout error only appears consistently during the final few iterations of the very last epoch. This makes me suspect that a data corruption issue might be less likely, as it would probably surface earlier or more randomly.

I did some further investigation and managed to trace the hang to the point right around the optimizer.step() call.

Given this specific behavior—stalling only at optimizer.step() at the very end of training—I was wondering if you had any further insights or advice on how to resolve it.

Thanks again for your time and help.

**Comment by jytime** [comment]
How about setting drop_last for the data sampler?

        self.sampler = DynamicDistributedSampler(self.dataset, seed=seed, shuffle=shuffle, drop_last=drop_last)

**Comment by zhangshuoneu** [comment]
Thanks, I'll try it.

**Comment by zyhbili** [comment]
I encounter the same problem. Have you solved it yet?

**Comment by zhangshuoneu** [comment]
Hi, @zyhbili
I believe I've found the cause of this issue. It seems you, like me, are running into this problem when training on multiple GPUs using only the CO3D dataset provided by jianyuan.

A direct solution is to either reduce the number of iterations per epoch (limit_train_batches) or increase the length of your dataset. The specific parameter can be found here in the configuration file:
https://github.com/facebookresearch/vggt/blob/8492456ce358ee9a4fe3274e36d73106b640fb5c/training/config/default.yaml#L13

Here is a more detailed explanation of the root cause:

Jianyuan's training code is set up to stop the current epoch once limit_train_batches is reached. However, in VGGT's multi-GPU dataloader setup, the number of iterations required for each process to finish an epoch is likely to be different. This is because the number of views per batch is dynamic to accommodate a variable number of input views, which in turn causes the dataset sampler's iteration count to vary across different GPU processes.

if the limit_train_batches parameter is bigger than the numeber of one  epoch iteration., in the final epoch, this leads to a race condition:
1. Some GPU processes finish iterating through their portion of the dataset early and stop updating gradients.
2. The processes that have not yet finished are stuck waiting for gradient broadcasts from the processes that have already terminated.
3. This wait eventually leads to a timeout, causing the program to crash with the error we are seeing.

By either reducing limit_train_batches or expanding the dataset, you can ensure all processes complete their iterations more uniformly, avoiding the timeout.

Hope this helps!

**Comment by WindVChen** [comment]
@zhangshuoneu  I also encountered the same problem, and I must say your answer was really insightful. It helped me a lot—thanks!

By the way, I think there is another solution that is to set the dataset length to a very large number, like 1e8, and enable the `inside_random` argument in the config. This way, we can ensure there are always enough samples to fetch.

**Comment by zyhbili** [comment]
@zhangshuoneu Thanks for your insightful answer! Currently, i run the training smoothly.
